# Week 5 Assignment - Part 5.7

Repeat 5.6 by using the dictionary approach: After getting the list from getArtListUserfile, you will create a Python dictionary that will look like the following: artdict = { artist1: 0, artist2: 0, artist3: 0, ..... artistn:0} where n is total number of artists (here zeros depict count initialized to 0s). Next you will read each (let us say ith artist) record in the file user_artist_data.txt and update the count for that particular artist by using code similar to the following:
artdict[i] = artdict[i] + tms 
where tms is the number in the last column.

On my Macbook Pro, this function takes less than a minute to execute.

In [1]:
import time

# Generator function to read a file - avoids memory issues
def read_file(filename,split_char=' '):
    with open(filename,mode='r',encoding="utf-8") as f:
        while True:
            try:
                file_line = next(f)
            except MemoryError as mem_error:
                print(f"Only {len(file_list)} records were read from '{filename}' due to memory constraints: {mem_error}")
            except StopIteration:
                break
            else:
                yield file_line.strip().split(split_char)

# Dictionaries are a better way to do this, and if we can loop once through the data even better
def get_most_played_artist():
    
    most = ['', 0]
    artists = {}
    
    for rec in read_file('user_artist_data.txt'):
        if rec[1] in artists:
            artists[rec[1]] += int(rec[2])
        else:
            artists[rec[1]] = int(rec[2])
            
        if artists[rec[1]] > most[1]:
            most = [rec[1], artists[rec[1]]]
#     print(f"Count: {counter} Artist: '{str(artist)}' Played: {artist_played:,}")
    return most


In [2]:
started = time.time()

most_played = get_most_played_artist()

elapsed = time.time() - started

print(f"Most played artist across all users was '{most_played[0]}' with {most_played[1]:,} plays.")
print(f"Time elapsed to find the most played artist: {elapsed:06.6f} seconds")

Most played artist across all users was '979' with 2,502,130 plays.
Time elapsed to find the most played artist: 27.849024 seconds


#### My Results

Artist '979' had over 2.5 millions total plays from all users.

It took 28 seconds to find this result on my laptop with Intel i7 and 16GB RAM.